metron-interface/metron-alerts/src/app/alerts/alerts-list/tree-view/tree-view.component.ts

```diff
@@ -0,0 +1,319 @@
+import {SearchResponse} from '../../../model/search-response';
+import {SearchService} from '../../../service/search.service';
+import {TreeGroupData, TreeAlertsSubscription} from './tree-group-data';
+import {GroupResponse} from '../../../model/group-response';
+import {GroupResult} from '../../../model/group-result';
+import {Group} from '../../../model/group';
+import {SortField} from '../../../model/sort-field';
+import {Sort} from '../../../utils/enums';
+import {MetronDialogBox, DialogType} from '../../../shared/metron-dialog-box';
+import {ElasticsearchUtils} from '../../../utils/elasticsearch-utils';
+import {SearchRequest} from '../../../model/search-request';
+
+@Component({
+  selector: 'app-tree-view',
+  templateUrl: './tree-view.component.html',
+  styleUrls: ['./tree-view.component.scss']
+})
+
+export class TreeViewComponent extends TableViewComponent implements OnChanges {
+
+  groupByFields: string[] = [];
+  topGroups: TreeGroupData[] = [];
+  groupResponse: GroupResponse = new GroupResponse();
+  treeGroupSubscriptionMap: {[key: string]: TreeAlertsSubscription } = {};
+
+  constructor(router: Router,
+              searchService: SearchService,
+              metronDialogBox: MetronDialogBox) {
+    super(router, searchService, metronDialogBox);
+  }
+
+  collapseGroup(groupArray: TreeGroupData[], level: number, index: number) {
+    for (let i = index + 1; i < groupArray.length; i++) {
+      if (groupArray[i].level > (level)) {
+        groupArray[i].show = false;
+        groupArray[i].expand = false;
+      } else {
+        break;
+      }
+    }
+  }
+
+  createQuery(selectedGroup: TreeGroupData) {
+    let searchQuery = this.queryBuilder.generateSelect();
+    let groupQery = Object.keys(selectedGroup.groupQueryMap).map(key => {
+      return key.replace(/:/g, '\\:') +
+          ':' +
+          String(selectedGroup.groupQueryMap[key])
+          .replace(/[\*\+\-=~><\"\?^\${}\(\)\:\!\/[\]\\\s]/g, '\\$&') // replace single  special characters
+          .replace(/\|\|/g, '\\||') // replace ||
+          .replace(/\&\&/g, '\\&&'); // replace &&
+    }).join(' AND ');
+
+    groupQery += searchQuery === '*' ? '' : (' AND ' + searchQuery);
+    return groupQery;
+  }
+
+  expandGroup(groupArray: TreeGroupData[], level: number, index: number) {
+    for (let i = index + 1; i < groupArray.length; i++) {
+      if (groupArray[i].level === (level + 1)) {
+        groupArray[i].show = true;
+      } else {
+        break;
+      }
+    }
+  }
+
+  getAlerts(selectedGroup: TreeGroupData): Subscription {
+    let searchRequest = new SearchRequest();
+    searchRequest.query = this.createQuery(selectedGroup);
+    searchRequest.from = selectedGroup.pagingData.from;
+    searchRequest.size = selectedGroup.pagingData.size;
+    searchRequest.sort = selectedGroup.sortField ? [selectedGroup.sortField] : [];
+
+    return this.searchGroup(selectedGroup, searchRequest);
```

Comment:
@iraghumitra I noticed that when grouping there are cases when the two locations where we display total counts will differ.  See the attached screenshots.  I underlined in yellow the counts that I am talking about.

![screen shot 2017-10-09 at 3 26 49 pm](https://user-images.githubusercontent.com/2475409/31356387-df5454f8-ad0b-11e7-9e0e-064e6b8694cd.png)

![screen shot 2017-10-09 at 3 23 09 pm](https://user-images.githubusercontent.com/2475409/31356388-df593b8a-ad0b-11e7-8c0b-58b21d58a73f.png)

It seems that the data used to build the total count in the group header is stale in comparison with the data used to display the individual alerts. We initially call `/group` to display the group header.  When a user expands the group we issue a `/search`, which retrieves fresher data.  Data that likely has a different total from what is currently displayed in the group header.

Can we update the count in the group header somewhere?  Maybe around here?

Reply:
@nickwallen that's a really nice catch, fixed it

Reply:
Do we have a test case for this?  I didn't see one in the commit that addressed this issue.

Reply:
I confirmed manually that the latest does indeed fix this issue.

@merrimanr does have a good point.  A quick unit test of this would be great.

Reply:
Actually, I spoke too soon.  @iraghumitra Your fix works great for a single group. But if you are grouping by multiple fields, we still get in a state where the group headers are showing stale data.  

In this screen shot, I am grouping by source-type, ip_src_addr, then country.  
* The count for the "country" group was correctly updated as soon as I expanded the header.
* The count for the "ip_src_addr" group was NOT updated.  It still displays the stale count of 1568.
* The count for the "source:type" group also was correctly updated.

![screen shot 2017-10-12 at 3 10 40 pm](https://user-images.githubusercontent.com/2475409/31514968-8d9da170-af61-11e7-970f-40ac0f8d4018.png)

Reply:
I am also noticing another issue with the group header counts that may or may not be related.  

In this example, I am grouping by 'source:type' then 'host'.  Since 'host' is specific to bro, there should be no alerts under the "snort" group.  If I expand the Snort group, I correctly see no alerts. 

The Snort group count is reported as 17,300 whereas I would expect that to be 0.  The count of 17,300 is tallying how many Snort alerts there are.  But adding a sub-group like 'host' can filter what the user will see when drilling down in the alerts.  

IMHO We should always expect the group count to reflect how many alerts the user will see when drilling down.  

![screen shot 2017-10-12 at 3 41 29 pm](https://user-images.githubusercontent.com/2475409/31516283-07d8d46a-af66-11e7-9829-ffa506e0c91c.png)

Reply:
@nickwallen  Again the problem is with how we are handling null values in the rest api.  The host name for all the alerts is emty so we are not seeing any alerts there.

Reply:
@iraghumitra That doesn't explain this one, right?  https://github.com/apache/metron/pull/768#discussion_r144387064

Reply:
I believe this issue is similar to https://github.com/apache/metron/pull/768#issuecomment-335281376 and should be resolved in the same follow on PR.  Do you agree @nickwallen?

Reply:
Yes, that makes sense to me. @merrimanr